killswitch.py

```diff
@@ -1,54 +1,231 @@
-
+OLD_KILLSWITCH_URL = 'https://raw.githubusercontent.com/EDCD/EDMarketConnector/releases/killswitches.json'
+DEFAULT_KILLSWITCH_URL = 'https://raw.githubusercontent.com/EDCD/EDMarketConnector/releases/killswitches_v2.json'
+CURRENT_KILLSWITCH_VERSION = 2
+UPDATABLE_DATA = Union[Mapping, Sequence]
 _current_version: semantic_version.Version = config.appversion_nobuild()
 
 
-class KillSwitch(NamedTuple):
+class SingleKill(NamedTuple):
+    """A single KillSwitch. Possibly with additional rules."""
+
+    match: str
+    reason: str
+    redact_fields: Optional[List[str]] = None
+    delete_fields: Optional[List[str]] = None
+    set_fields: Optional[Dict[str, Any]] = None
+
+    @property
+    def has_rules(self) -> bool:
+        """Return whether or not this SingleKill can apply rules to a dict to make it safe to use."""
+        return any(x is not None for x in (self.redact_fields, self.delete_fields, self.set_fields))
+
+    def apply_rules(self, target: UPDATABLE_DATA) -> UPDATABLE_DATA:
+        """
+        Apply the rules this SingleKill instance has to make some data okay to send.
+
+        Note that this MODIFIES DATA IN PLACE.
+
+        :param target: data to apply a rule to
+        """
+        # fields that contain . might be going deeper into the dict. check here FIRST to see if they exist at the top
+        # level, if not, then work our way down
+
+        for key in (self.redact_fields if self.redact_fields is not None else []):
+            _deep_apply(target, key, "REDACTED")
+
+        for key in (self.delete_fields if self.delete_fields is not None else []):
+            _deep_apply(target, key, delete=True)
+
+        for key, value in (self.set_fields if self .set_fields is not None else {}).items():
+            _deep_apply(target, key, value)
+
+        return target
+
+
+def _apply(target: UPDATABLE_DATA, key: str, to_set: Any = None, delete: bool = False):
+    """
+    Set or delete the given target key on the given target.
+
+    :param target: The thing to set data on
+    :param key: the key or index to set the data to
+    :param to_set: the data to set, if any, defaults to None
+    :param delete: whether or not to delete the key or index, defaults to False
+    :raises ValueError: when an unexpected target type is passed
+    """
+    if isinstance(target, MutableMapping):
+        if delete:
+            target.pop(key, None)
+        else:
+            target[key] = to_set
+
+    elif isinstance(target, MutableSequence):
+        idx = _get_int(key)
+        if idx is None:
+            raise ValueError(f'Cannot use string {key!r} as int for index into Sequence')
+
+        if delete:
+            if len(target) > idx:
+                target.pop(idx)
+
+        elif len(target) == idx:
+            target.append(to_set)
+
+        else:
+            target[idx] = to_set  # this can raise, that's fine
```

Comment:
Negative index also an issue here.

We could decide to just not support them, but I could see the usefulness given it's valid python slicing syntax.

Reply:
1. Either this needs to allow for setting a non-existent index that isn't an append.

2. OR Killswitches.md needs to document that an index on a set must either be an existing one, or only a simple append.

3. OR this function needs to be marked as explicitly possibly raising `IndexError`.

2 or 3 needs a test put in to ensure that happens as documented.

I added:

```python
        (['test append'], '2', '', 'yay', ['test append', 'yay']),
```
to the 'test_apply()' parameters in `test_apply.py` and it blew up with an IndexError.